<h1>Capstone Project</h1>

<p>This is a notebook for my Coursera Applied Data Science Capstone Project</p>

<h2>Objective</h2>

<p>Gathering valuable information and insights on car accidents in Seattle. Then using this information to create a machine learning model that is capable of predicting the severity of traffic accidents based on certain attributes.</p>

<font size="+1"><b>Importing Libraries</b></font>

In [14]:
import pandas as pd
import numpy as np

<h2>Load the Data</h2>

In [15]:
df = pd.read_csv(r'Data-Collisions.csv')

C:\Users\fredr\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<h2>Data Exploration and Cleaning</h2>

In [16]:
pd.set_option("max_columns", 50)
df.head()

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  \
0  Matched  Intersection  37475.0   
1  Matched         Block      NaN   
2  Matched         Block      NaN   
3  Matched         Block      NaN   
4  Matched  Intersection  34387.0   

                                      LOCATION EXCEPTRSNCODE EXCEPTRSNDESC  \
0                   5TH AVE NE AND NE 103RD ST                         NaN   
1   AURORA BR BETWEEN RAYE ST AND BRIDGE WAY N           NaN           NaN   
2  4TH AVE BETWEEN SENECA ST AND UNIVERSITY ST           NaN           NaN   
3     2ND AVE BETWEEN MARION ST AND MADISON ST                         NaN   
4              SWIFT AVE S AND SWIFT AV OFF RP           NaN           NaN   

   SEVERITYCODE.1                    SEVERITYDESC COLLISIONTYPE  PERSONCOUNT  \
0               2                Injury Collision        Angles            2   
1               1  Property Damage Only Collision     Sideswipe            2   
2               1  Property Damage Only Collision    Parked Car            4   
3               1  Property Damage Only Collision         Other            3   
4               2                Injury Collision        Angles            2   

   PEDCOUNT  PEDCYLCOUNT  VEHCOUNT                 INCDATE  \
0         0            0         2  2013/03/27 00:00:00+00   
1         0            0         2  2006/12/20 00:00:00+00   
2         0            0         3  2004/11/18 00:00:00+00   
3         0            0         3  2013/03/29 00:00:00+00   
4         0            0         2  2004/01/28 00:00:00+00   

                  INCDTTM                             JUNCTIONTYPE  \
0    3/27/2013 2:54:00 PM   At Intersection (intersection related)   
1   12/20/2006 6:55:00 PM  Mid-Block (not related to intersection)   
2  11/18/2004 10:20:00 AM  Mid-Block (not related to intersection)   
3    3/29/2013 9:26:00 AM  Mid-Block (not related to intersection)   
4    1/28/2004 8:04:00 AM   At Intersection (intersection related)   

   SDOT_COLCODE                                       SDOT_COLDESC  \
0            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
1            16  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...   
2            14       MOTOR VEHICLE STRUCK MOTOR VEHICLE, REAR END   
3            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
4            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   

  INATTENTIONIND UNDERINFL   WEATHER ROADCOND                LIGHTCOND  \
0            NaN         N  Overcast      Wet                 Daylight   
1            NaN         0   Raining      Wet  Dark - Street Lights On   
2            NaN         0  Overcast      Dry                 Daylight   
3            NaN         N     Clear      Dry                 Daylight   
4            NaN         0   Raining      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Enterin

 <font size="+1"><b>Rows and Columns</b></font>

In [17]:
df.shape

(194673, 38)

<p><b>The data set contains 38 columns (severitycode is duplicated) and 194673 rows.</b></p>

 <font size="+1"><b>Timeframe</b></font>

In [18]:
df.sort_values(by='INCDTTM')[['INCDTTM', 'INCDATE']].head()

INCDTTM                 INCDATE
7996   1/1/2004  2004/01/01 00:00:00+00
10714  1/1/2004  2004/01/01 00:00:00+00
5083   1/1/2004  2004/01/01 00:00:00+00
3291   1/1/2004  2004/01/01 00:00:00+00
5940   1/1/2004  2004/01/01 00:00:00+00

<b>From 1/1/2004</b>

In [19]:
df.sort_values(by='INCDTTM', ascending=False)[['INCDTTM', 'INCDATE']].head()

INCDTTM                 INCDATE
192339  9/9/2019 9:27:00 AM  2019/09/09 00:00:00+00
191820  9/9/2019 9:10:00 PM  2019/09/09 00:00:00+00
191832  9/9/2019 9:08:00 AM  2019/09/09 00:00:00+00
192599  9/9/2019 9:06:00 AM  2019/09/09 00:00:00+00
190713  9/9/2019 8:45:00 AM  2019/09/09 00:00:00+00

<b>to 9/9/2019</b>

<h3>Data Types</h3>

In [20]:
df.dtypes

SEVERITYCODE        int64
X                 float64
Y                 float64
OBJECTID            int64
INCKEY              int64
COLDETKEY           int64
REPORTNO           object
STATUS             object
ADDRTYPE           object
INTKEY            float64
LOCATION           object
EXCEPTRSNCODE      object
EXCEPTRSNDESC      object
SEVERITYCODE.1      int64
SEVERITYDESC       object
COLLISIONTYPE      object
PERSONCOUNT         int64
PEDCOUNT            int64
PEDCYLCOUNT         int64
VEHCOUNT            int64
INCDATE            object
INCDTTM            object
JUNCTIONTYPE       object
SDOT_COLCODE        int64
SDOT_COLDESC       object
INATTENTIONIND     object
UNDERINFL          object
WEATHER            object
ROADCOND           object
LIGHTCOND          object
PEDROWNOTGRNT      object
SDOTCOLNUM        float64
SPEEDING           object
ST_COLCODE         object
ST_COLDESC         object
SEGLANEKEY          int64
CROSSWALKKEY        int64
HITPARKEDCAR       object
dtype: objec

<h4>Cast to datetime</h4>

<b>The type of the INCDATE and INCDTTM needs to be cast to datetime, and we only need one of the two column.</b>

In [21]:
df = df.astype({'INCDTTM': 'datetime64'})
df.drop(['INCDATE'], axis=1, inplace=True)
df.head()

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  \
0  Matched  Intersection  37475.0   
1  Matched         Block      NaN   
2  Matched         Block      NaN   
3  Matched         Block      NaN   
4  Matched  Intersection  34387.0   

                                      LOCATION EXCEPTRSNCODE EXCEPTRSNDESC  \
0                   5TH AVE NE AND NE 103RD ST                         NaN   
1   AURORA BR BETWEEN RAYE ST AND BRIDGE WAY N           NaN           NaN   
2  4TH AVE BETWEEN SENECA ST AND UNIVERSITY ST           NaN           NaN   
3     2ND AVE BETWEEN MARION ST AND MADISON ST                         NaN   
4              SWIFT AVE S AND SWIFT AV OFF RP           NaN           NaN   

   SEVERITYCODE.1                    SEVERITYDESC COLLISIONTYPE  PERSONCOUNT  \
0               2                Injury Collision        Angles            2   
1               1  Property Damage Only Collision     Sideswipe            2   
2               1  Property Damage Only Collision    Parked Car            4   
3               1  Property Damage Only Collision         Other            3   
4               2                Injury Collision        Angles            2   

   PEDCOUNT  PEDCYLCOUNT  VEHCOUNT             INCDTTM  \
0         0            0         2 2013-03-27 14:54:00   
1         0            0         2 2006-12-20 18:55:00   
2         0            0         3 2004-11-18 10:20:00   
3         0            0         3 2013-03-29 09:26:00   
4         0            0         2 2004-01-28 08:04:00   

                              JUNCTIONTYPE  SDOT_COLCODE  \
0   At Intersection (intersection related)            11   
1  Mid-Block (not related to intersection)            16   
2  Mid-Block (not related to intersection)            14   
3  Mid-Block (not related to intersection)            11   
4   At Intersection (intersection related)            11   

                                        SDOT_COLDESC INATTENTIONIND UNDERINFL  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         N   
1  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...            NaN         0   
2       MOTOR VEHICLE STRUCK MOTOR VEHICLE, REAR END            NaN         0   
3  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         N   
4  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         0   

    WEATHER ROADCOND                LIGHTCOND PEDROWNOTGRNT  SDOTCOLNUM  \
0  Overcast      Wet                 Daylight           NaN         NaN   
1   Raining      Wet  Dark - Street Lights On           NaN   6354039.0   
2  Overcast      Dry                 Daylight           NaN   4323031.0   
3     Clear      Dry                 Daylight           NaN         NaN   
4   Raining      Wet                 Daylight           NaN   4028032.0   

  SPEEDING ST_COLCODE                                         ST_COLDESC  \
0      NaN         10                                  Entering at angle   
1      NaN         11  From same direction - both going straight - bo...   
2      NaN         32                             One parked--one moving   
3      NaN         23                   From same direction - all others   
4      NaN         10                                  Entering at angle   

   SEGLANEKEY  CROSSWALKKEY HITPARKEDCAR  
0           0             0            N  
1           0             0            N  
2           0             0            N  
3           0             0            N  
4           0             0  

<h3>Removing obvious columns</h3>

<b>X and Y columns contains the coordinates for the incident. We can drop these, though we will be using the LOCATION column for the location of the accident.</b>

In [30]:
df.drop(['X', 'Y'], axis=1, inplace=True)
df.head(1)

SEVERITYCODE  OBJECTID  INCKEY  COLDETKEY REPORTNO   STATUS      ADDRTYPE  \
0             2         1    1307       1307  3502005  Matched  Intersection   

    INTKEY                    LOCATION EXCEPTRSNCODE EXCEPTRSNDESC  \
0  37475.0  5TH AVE NE AND NE 103RD ST                         NaN   

   SEVERITYCODE.1      SEVERITYDESC COLLISIONTYPE  PERSONCOUNT  PEDCOUNT  \
0               2  Injury Collision        Angles            2         0   

   PEDCYLCOUNT  VEHCOUNT             INCDTTM  \
0            0         2 2013-03-27 14:54:00   

                             JUNCTIONTYPE  SDOT_COLCODE  \
0  At Intersection (intersection related)            11   

                                        SDOT_COLDESC INATTENTIONIND UNDERINFL  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         N   

    WEATHER ROADCOND LIGHTCOND PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0  Overcast      Wet  Daylight           NaN         NaN      NaN         10   

          ST_COLDESC  SEGLANEKEY  CROSSWALKKEY HITPARKEDCAR  
0  Entering at angle           0             0            N

<b>There's no information on the EXCEPTRSNCODE and EXCEPTRSNDESC columns in the metadata.</b>

In [31]:
df.drop(['EXCEPTRSNCODE', 'EXCEPTRSNDESC'], axis=1, inplace=True)
df.head(1)

SEVERITYCODE  OBJECTID  INCKEY  COLDETKEY REPORTNO   STATUS      ADDRTYPE  \
0             2         1    1307       1307  3502005  Matched  Intersection   

    INTKEY                    LOCATION  SEVERITYCODE.1      SEVERITYDESC  \
0  37475.0  5TH AVE NE AND NE 103RD ST               2  Injury Collision   

  COLLISIONTYPE  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  VEHCOUNT  \
0        Angles            2         0            0         2   

              INCDTTM                            JUNCTIONTYPE  SDOT_COLCODE  \
0 2013-03-27 14:54:00  At Intersection (intersection related)            11   

                                        SDOT_COLDESC INATTENTIONIND UNDERINFL  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         N   

    WEATHER ROADCOND LIGHTCOND PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0  Overcast      Wet  Daylight           NaN         NaN      NaN         10   

          ST_COLDESC  SEGLANEKEY  CROSSWALKKEY HITPARKEDCAR  
0  Entering at angle           0             0            N

<b>The INTKEY column contains the key that corresponds to the intersection associated with a collision. Since ADDRTYPE displays the type of address (alley, block or intersection) and LOCATION specifies the location, we can drop the INTKEY columns<b>

In [32]:
df.drop(['INTKEY'], axis=1, inplace=True)
df.head(1)

SEVERITYCODE  OBJECTID  INCKEY  COLDETKEY REPORTNO   STATUS      ADDRTYPE  \
0             2         1    1307       1307  3502005  Matched  Intersection   

                     LOCATION  SEVERITYCODE.1      SEVERITYDESC COLLISIONTYPE  \
0  5TH AVE NE AND NE 103RD ST               2  Injury Collision        Angles   

   PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  VEHCOUNT             INCDTTM  \
0            2         0            0         2 2013-03-27 14:54:00   

                             JUNCTIONTYPE  SDOT_COLCODE  \
0  At Intersection (intersection related)            11   

                                        SDOT_COLDESC INATTENTIONIND UNDERINFL  \
0  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...            NaN         N   

    WEATHER ROADCOND LIGHTCOND PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0  Overcast      Wet  Daylight           NaN         NaN      NaN         10   

          ST_COLDESC  SEGLANEKEY  CROSSWALKKEY HITPARKEDCAR  
0  Entering at angle           0             0            N

<h3>Missing Values</h3>

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194673 entries, 0 to 194672
Data columns (total 37 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   SEVERITYCODE    194673 non-null  int64         
 1   X               189339 non-null  float64       
 2   Y               189339 non-null  float64       
 3   OBJECTID        194673 non-null  int64         
 4   INCKEY          194673 non-null  int64         
 5   COLDETKEY       194673 non-null  int64         
 6   REPORTNO        194673 non-null  object        
 7   STATUS          194673 non-null  object        
 8   ADDRTYPE        192747 non-null  object        
 9   INTKEY          65070 non-null   float64       
 10  LOCATION        191996 non-null  object        
 11  EXCEPTRSNCODE   84811 non-null   object        
 12  EXCEPTRSNDESC   5638 non-null    object        
 13  SEVERITYCODE.1  194673 non-null  int64         
 14  SEVERITYDESC    194673 non-null  obj

<b>There are clearly a lot of missing values in several of the columns</b>

In [28]:
df.isnull().sum()

SEVERITYCODE           0
X                   5334
Y                   5334
OBJECTID               0
INCKEY                 0
COLDETKEY              0
REPORTNO               0
STATUS                 0
ADDRTYPE            1926
INTKEY            129603
LOCATION            2677
EXCEPTRSNCODE     109862
EXCEPTRSNDESC     189035
SEVERITYCODE.1         0
SEVERITYDESC           0
COLLISIONTYPE       4904
PERSONCOUNT            0
PEDCOUNT               0
PEDCYLCOUNT            0
VEHCOUNT               0
INCDTTM                0
JUNCTIONTYPE        6329
SDOT_COLCODE           0
SDOT_COLDESC           0
INATTENTIONIND    164868
UNDERINFL           4884
WEATHER             5081
ROADCOND            5012
LIGHTCOND           5170
PEDROWNOTGRNT     190006
SDOTCOLNUM         79737
SPEEDING          185340
ST_COLCODE            18
ST_COLDESC          4904
SEGLANEKEY             0
CROSSWALKKEY           0
HITPARKEDCAR           0
dtype: int64

<b>Columns containig missing values</b>

In [29]:
df.columns[df.isnull().any()]

Index(['X', 'Y', 'ADDRTYPE', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE',
       'EXCEPTRSNDESC', 'COLLISIONTYPE', 'JUNCTIONTYPE', 'INATTENTIONIND',
       'UNDERINFL', 'WEATHER', 'ROADCOND', 'LIGHTCOND', 'PEDROWNOTGRNT',
       'SDOTCOLNUM', 'SPEEDING', 'ST_COLCODE', 'ST_COLDESC'],
      dtype='object')

<h4>Handling Missing values</h4>

<b>Dropping column which won't be used in the further analysis:</b>

<font size="+1"><b>Descriptive analysis</b></font>

In [10]:
df.describe()

SEVERITYCODE              X              Y       OBJECTID  \
count  194673.000000  189339.000000  189339.000000  194673.000000   
mean        1.298901    -122.330518      47.619543  108479.364930   
std         0.457778       0.029976       0.056157   62649.722558   
min         1.000000    -122.419091      47.495573       1.000000   
25%         1.000000    -122.348673      47.575956   54267.000000   
50%         1.000000    -122.330224      47.615369  106912.000000   
75%         2.000000    -122.311937      47.663664  162272.000000   
max         2.000000    -122.238949      47.734142  219547.000000   

              INCKEY      COLDETKEY         INTKEY  SEVERITYCODE.1  \
count  194673.000000  194673.000000   65070.000000   194673.000000   
mean   141091.456350  141298.811381   37558.450576        1.298901   
std     86634.402737   86986.542110   51745.990273        0.457778   
min      1001.000000    1001.000000   23807.000000        1.000000   
25%     70383.000000   70383.000000   28667.000000        1.000000   
50%    123363.000000  123363.000000   29973.000000        1.000000   
75%    203319.000000  203459.000000   33973.000000        2.000000   
max    331454.000000  332954.000000  757580.000000        2.000000   

         PERSONCOUNT       PEDCOUNT    PEDCYLCOUNT       VEHCOUNT  \
count  194673.000000  194673.000000  194673.000000  194673.000000   
mean        2.444427       0.037139       0.028391       1.920780   
std         1.345929       0.198150       0.167413       0.631047   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000       0.000000       0.000000       2.000000   
50%         2.000000       0.000000       0.000000       2.000000   
75%         3.000000       0.000000       0.000000       2.000000   
max        81.000000       6.000000       2.000000      12.000000   

        SDOT_COLCODE    SDOTCOLNUM     SEGLANEKEY  CROSSWALKKEY  
count  194673.000000  1.149360e+05  194673.000000  1.946730e+05  
mean       13.867768  7.972521e+06     269.401114  9.782452e+03  
std         6.868755  2.553533e+06    3315.776055  7.226926e+04  
min         0.000000  1.007024e+06       0.000000  0.000000e+00  
25%        11.000000  6.040015e+06       0.000000  0.000000e+00  
50%        13.000000  8.023022e+06       0.000000  0.000000e+00  
75%        14.000000  1.015501e+07       0.000000  0.000000e+00  
max        69.000000  1.307202e+07  525241.000000  5.239700e+06

In [10]:
df.describe(include=['O'])

REPORTNO   STATUS ADDRTYPE  \
count    194673   194673   192747   
unique   194670        2        3   
top     1782439  Matched    Block   
freq          2   189786   126926   

                                                 LOCATION EXCEPTRSNCODE  \
count                                              191996         84811   
unique                                              24102             2   
top     BATTERY ST TUNNEL NB BETWEEN ALASKAN WY VI NB ...                 
freq                                                  276         79173   

                                            EXCEPTRSNDESC  \
count                                                5638   
unique                                                  1   
top     Not Enough Information, or Insufficient Locati...   
freq                                                 5638   

                          SEVERITYDESC COLLISIONTYPE                 INCDATE  \
count                           194673        189769                  194673   
unique                               2            10                    5985   
top     Property Damage Only Collision    Parked Car  2006/11/02 00:00:00+00   
freq                            136485         47987                      96   

          INCDTTM                             JUNCTIONTYPE  \
count      194673                                   188344   
unique     162058                                        7   
top     11/2/2006  Mid-Block (not related to intersection)   
freq           96                                    89800   

                                             SDOT_COLDESC INATTENTIONIND  \
count                                              194673          29805   
unique                                                 39              1   
top     MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...              Y   
freq                                                85209          29805   

       UNDERINFL WEATHER ROADCOND LIGHTCOND PEDROWNOTGRNT SPEEDING ST_COLCODE  \
count     189789  189592   189661    189503          4667     9333     194655   
unique         4      11        9         9             1        1        115   
top            N   Clear      Dry  Daylight             Y        Y         32   
freq      100274  111135   124510    116137          4667     9333      27612   

                    ST_COLDESC HITPARKEDCAR  
count                   189769       194673  
unique                      62            2  
top     One parked--one moving            N  
freq                     44421       187457

<p><b>From the count we can see that there are many columns with missing data.</b></p>